### PR TITLE
List v2: add forward delete

### PR DESCRIPTION
**Pull request by ellatrix**
## What?

Adds the ability to forward delete from the end of a list item. If the next line can be outdented, it will first be outdented. If it's at the top level, it will be merged.

forward-delete

forward-delete

## Why?

Feature parity with List v1.

## Testing Instructions

Try forward delete (Fn+Backspace) from various positions.

### CODE REVIEW

File: packages/block-library/src/list-item/hooks/use-outdent-list-item.js [hunk 1]
```diff
@@ -29,93 +30,90 @@ export default function useOutdentListItem( clientId ) {
...
 			);
 
-			// Replace the parent list item block, with a new block containing
-			// the previous siblings before the first block being outdent,
-			// followed by the blocks being outdent with the after siblings added
-			// as children of the last block.
-			replaceBlocks(
-				[ listItemParentId ],
-				[ newListItemParent, ...newBlocksExcludingLast, newLastItem ]
-			);
```
Comment (ellatrix) [inline packages/block-library/src/list-item/hooks/use-outdent-list-item.js:101]: @jorgefilipecosta @ntsekouras I cleaned up outdenting to use `moveBlocksToPosition`. This has the benefit of keeping block client IDs and having to reapply selection. It's also a lot simpler and easier to follow.

File: packages/block-library/src/list-item/edit.js [hunk 2]
```diff
@@ -42,14 +41,14 @@ function IndentUI( { clientId } ) {
 				title={ __( 'Outdent' ) }
 				describedBy={ __( 'Outdent list item' ) }
 				disabled={ ! canOutdent }
-				onClick={ outdentListItem }
+				onClick={ () => outdentListItem() }
```
Comment (jorgefilipecosta) [inline packages/block-library/src/list-item/edit.js:44]: Why are we doing this change here? Can't we just pass outdentListItem as before?

File: packages/block-library/src/list-item/hooks/use-outdent-list-item.js [hunk 3]
```diff
@@ -29,93 +30,90 @@ export default function useOutdentListItem( clientId ) {
...
-				.map( ( _clientId ) => cloneBlock( getBlock( _clientId ) ) );
-
-			// Create a new list item block whose attributes are equal to the
-			// last block being outdent and whose children are the children that it had (if any)
-			// followed by the siblings that existed after it.
-			const newLastItem = createListItem(
-				lastBlock.attributes,
-				hasChildItems ? childList.attributes : listAttributes,
-				[ ...childItems, ...afterSiblings ]
+			const parentList = getBlockRootClientId( firstClientId );
```
Comment (jorgefilipecosta) [inline packages/block-library/src/list-item/hooks/use-outdent-list-item.js]: I guess we could name this parentListId to make it clear that it is an id and not the block itself.

File: packages/block-library/src/list-item/hooks/use-outdent-list-item.js [hunk 4]
```diff
@@ -29,93 +30,90 @@ export default function useOutdentListItem( clientId ) {
...
+	}
+
 	return [
 		canOutdent,
-		useCallback( () => {
-			const _hasMultiSelection = hasMultiSelection();
-			const clientIds = _hasMultiSelection
-				? getMultiSelectedBlockClientIds()
-				: [ clientId ];
+		useCallback( ( clientIds = getSelectedBlockClientIds() ) => {
```
Comment (jorgefilipecosta) [inline packages/block-library/src/list-item/hooks/use-outdent-list-item.js:59]: The hook useOutdentListItem( clientId ) already receives a clientId parameter. It feels strange that the callback the hook returns ignores the parameter and receives a similar parameter can we just make the hook receive clientIds defaulting to getSelectedBlockClientIds and keep the callback without parameters?

**Review by jorgefilipecosta — APPROVED** [review]
Left some comments/questions, but everything works great. Awesome work, @ellatrix 👍

File: packages/block-library/src/list-item/hooks/use-outdent-list-item.js (hunk 4 above)
Comment (ellatrix) [reply packages/block-library/src/list-item/hooks/use-outdent-list-item.js:59]: Not sure I'm following here. The default is to outdent the selected blocks (`clientIds = getSelectedBlockClientIds()`), but we also need the ability to outdent specific blocks for forward Delete (outdent the next block).

**Pull request merged**